### ISSUE TITLE
ws-auth apiServerZoneDetail does not handle domains being deleted gracefully

### ISSUE THREAD

**Issue by RobinGeuze** [PowerDNS/pdns#10100]
- Program: Authoritative
 - Issue type: Bug report

### Short description

There is a possibility for a race condition in the ws-auth code (which I encountered). Since domain info is retrieved outside of a transaction (out of necessity) this means that for example LMDB fails to create a transaction when calling startTransaction, which at the moment is not checked in the function. In my case this ended up causing a segfault, which should at least be "fixed" by #10099 turning it into a normal exception.

### Environment

 - Operating system: Debian Buster
 - Software version: 4.4.1
 - Software source: PowerDNS repository

### Steps to reproduce

Reproduction is rather hard, easiest way is probably to introduce a sleep in the function after retrieving and then using pdnsutil or some other means to delete the domain while the function sleeps.

### Expected behaviour

Either some error, or in the case of a DELETE potentially handling it gracefully (since the domain is already gone anyway).

### Actual behaviour
Without #10099 when using the LMDB backend it results in a segfault.

**Comment by zeha** [comment]
Do you have a backtrace or something? Right now I wouldn't know where to start looking.

`apiServerZoneDetail` does a lot of things - was it while querying, deleting, creating?

**Comment by RobinGeuze** [comment]
So the problem is that at https://github.com/PowerDNS/pdns/blob/cec701c1f2881059c5bd1359d3720c555c7c1a3d/pdns/ws-auth.cc#L1775 a domaininfo is retrieved to figure out which backend a domain is in. Then after that operation X is executed on that domain. However, since (obviously) there has been no transaction started when you retrieved the domain (since you don't know which backend to start it for), this means that by the time you do startTransaction in either the PUT or the DELETE clause the domain might have already been deleted. This used to cause a segfault in at least the LMDB backend but that was fixed.

**Comment by zeha** [comment]
I see. That sounds like pdnsutil would also crash under the same circumstances. I think the current UeberBackend + DNSBackend design basically requires backends to deal with this problem themselves, especially those backends using the domain_id in startTransaction. I don't think we should workaround this in the webserver; if it ought to be fixed, it should be a) either with a better UeberBackend/DNSBackend design or b) backends should conform to a uniform API and possibly throw a uniform exception ...

**Comment by RobinGeuze** [comment]
Agreed, possibly throwing a uniform exception would make the most sense.
